### ISSUE TITLE
Option to exclude certain buildings when calculating PP

### ISSUE THREAD

**Issue by JustALittleCrazyTalk** [Aktanusa/CookieMonster#155]
Because max mana effects the total time it takes to refill the mana bar I'm currently sitting at just 7 Wizard towers to maximize my baked goods casts.  This means that the building showing green is always wizard towers.  An optional flag to exclude them would restore functionality when pursuing this strategy.

**Comment by Aktanusa** [comment]
Hmm, I see what you mean.  How to implement it might not be too easy though.  We'll see.

**Comment by Kidsheldon** [comment]
I think making a toggle list of exclusions from calculations would work. You could toggle on/off buildings and upgrades, covering any niche cases like this.

**Comment by vinzzz001** [comment]
Something like this?

**Comment by Aktanusa** [comment]
making it gray is the easy part, the hard part is making the UI so people can choose what to place on the ignore list =p.  I think a better way is to rearranging the buildings from best to worse PP so then you can always ignore the first building or whatever.  It also makes clicking the best PP building easier with it always on top.

**Comment by dhouck** [comment]
I don't want the buildings to be rearranged; I like having them in order of "best" to "worst" (even though this order is only approximate because of different upgrades and especially synergies). I'm also used to finding them in the default order. It would be nice to make the Wizard Tower (or whatever building) gray without affecting everything else.

Since the Wizard Tower is more likely to be on the list than anything else, a short-term solution might be to make a toggle for it specifically in the options screen.

**Comment by Aktanusa** [comment]
My calculations actually include synergies, so it's not really approximate, but pretty close to the most efficient order.

**Comment by dhouck** [comment]
No, I mean having the buildings in cursor-grandma-farm order, which is approximately worst to best CPS/building.

**Comment by Aktanusa** [comment]
I'm still confused.  I'll reread and try to make sense of it.

**Comment by dhouck** [comment]
Sorry, sometimes I can explain things poorly. Let me try again; hopefully I'll phrase it better now.

If I read correctly, you suggested reordering the buildings based on PP instead of adjusting which ones are included in the color calculations. I don't think this is the best approach; I prefer to keep the buildings in default order.

Another objection is that reordering buildings might help for finding the second-best building, but it would prevent upgrades from being colored "better than any building" unless they were better than the *wizard tower* or other ignored buildings.

**Comment by Aktanusa** [comment]
I got that, what confused me was this quote: " (even though this order is only approximate because of different upgrades and especially synergies)".  I am saying the calculations take synergies into account.

**Comment by dhouck** [comment]
That was referring to the default order. In a first or Born Again run, the default order is the order of increasing benefit for having one of those buildings, except when upgrades and synergies change the benefits. It is also the order of increasing price and of decreasing PP if you own the
Owning upgrades or synergies changes what order is the order of increasing benefit for buildings, but

**Comment by ctvf143** [comment]
so, is there anything i can do with that isuue?

**Comment by dhouck** [comment]
The discussion got off track, but I think a way to toggle buildings to not be included in the best building calculation would be good. The colors are almost useless if you’re actively playing and using mana, because then wizard towers will just always be the green building and none of the upgrades will go cyan.

**Comment by dhouck** [comment]
I’m currently using Chrome overrides to keep a local version, with these changes:

```diff
--- online/CookieMonster.js
+++ local/CookieMonster.js
@@ -92,6 +92,9 @@
 	for (var i in CM.Cache.Objects) {
 		//CM.Cache.Objects[i].pp = Game.Objects[i].getPrice() / CM.Cache.Objects[i].bonus;
 		CM.Cache.Objects[i].pp = (Math.max(Game.Objects[i].getPrice() - (Game.cookies + CM.Disp.GetWrinkConfigBank()), 0) / Game.cookiesPs) + (Game.Objects[i].getPrice() / CM.Cache.Objects[i].bonus);
+		if (i == 'Wizard tower') {
+			continue;
+		}
 		if (CM.Cache.min == -1 || CM.Cache.Objects[i].pp < CM.Cache.min) CM.Cache.min = CM.Cache.Objects[i].pp;
 		if (CM.Cache.max == -1 || CM.Cache.Objects[i].pp > CM.Cache.max) CM.Cache.max = CM.Cache.Objects[i].pp;
 	}
@@ -100,6 +103,7 @@
 		var color = '';
 		if (CM.Cache.Objects[i].pp == CM.Cache.min) color = CM.Disp.colorGreen;
 		else if (CM.Cache.Objects[i].pp == CM.Cache.max) color = CM.Disp.colorRed;
+		else if (i == 'Wizard tower') color = CM.Disp.colorGray;
 		else if (CM.Cache.Objects[i].pp > CM.Cache.mid) color = CM.Disp.colorOrange;
 		else color = CM.Disp.colorYellow;
 		CM.Cache.Objects[i].color = color;
@@ -128,7 +132,7 @@
 		//CM.Cache[target][i].pp = CM.Cache[target][i].price / CM.Cache[target][i].bonus;
 		CM.Cache[target][i].pp = (Math.max(CM.Cache[target][i].price - (Game.cookies + CM.Disp.GetWrinkConfigBank()), 0) / Game.cookiesPs) + (CM.Cache[target][i].price / CM.Cache[target][i].bonus);
 		var color = '';
-		if (CM.Cache[target][i].pp <= 0 || CM.Cache[target][i].pp == Infinity) color = CM.Disp.colorGray;
+		if (CM.Cache[target][i].pp <= 0 || CM.Cache[target][i].pp == Infinity || i == 'Wizard tower') color = CM.Disp.colorGray;
 		else if (CM.Cache[target][i].pp < CM.Cache.min) color = CM.Disp.colorBlue;
 		else if (CM.Cache[target][i].pp == CM.Cache.min) color = CM.Disp.colorGreen;
 		else if (CM.Cache[target][i].pp == CM.Cache.max) color = CM.Disp.colorRed;
```

Obviously this isn’t a final solution because it should consult preferences.  It’s also debatable whether the color should actually be gray or if it should range from blue to purple as appropriate; maybe that should also be a preference.

**Comment by Aktanusa** [comment]
Thanks for helping out as I've not been setting this addon as a priority for a while now.  Maybe that will change in the future.
